compaction_picker.go

```diff
@@ -111,27 +107,6 @@ func (p *compactionPicker) initLevelMaxBytes(v *version, opts *db.Options) {
 		}
 	}
 
-	if l0Size > baseLevelSize &&
-		(l0Size > opts.L1MaxBytes ||
-			(len(v.files)/2) >= opts.L0CompactionThreshold) {
-		// We adjust the base level according to actual L0 size, and adjust the
-		// level multiplier accordingly, when:
-		//
-		//   1. the L0 size is larger than level size base, or
-		//   2. number of L0 files reaches twice the L0->L1 compaction threshold
-		//
-		// We don't do this otherwise to keep the LSM-tree structure stable unless
-		// the L0 compaction is backlogged.
-		baseLevelSize = l0Size
-		if p.baseLevel == numLevels-1 {
-			levelMultiplier = 1.0
-		} else {
-			levelMultiplier = math.Pow(
```

Comment:
This one's interesting because it smooths the LSM fanout. For example if you have 2000MB in L6, and base level L4 with 100MB. Without this change you'll have:

- L4 100MB
- L5 200MB (fanout 2)
- L6 2000MB (fanout 10)
- Write-amp to compact from L4 to L6 is 2+10=12

With this change you'll have:

- L4 100MB
- L5 450MB (fanout 4.5)
- L6 450MB (fanout 4.5)
- Write-amp to compact from L4 to L6 is 4.5+4.5=9

Like a lot of things in RocksDB, it was only enabled in unusual edge cases (here, during write-burst mode when we've all but given up on space-amp) to minimize chance of impacting existing use cases. But with Pebble, I wonder if things like this are good to enable by default.

Reply:
Mildly interesting side note: the first LSM paper (http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.44.2782&rep=rep1&type=pdf) proved this result about smoothing fanout to minimize write-amp in Section 3.4.

Reply:
> But with Pebble, I wonder if things like this are good to enable by default.

Maybe. It isn't obvious to me that this result holds if the amount of data in the DB is changing.

I've been thinking more about the compaction behavior in Pebble for the `sync` workload (aka `kv0`) and have found a badness that I think that applies to RocksDB. I'm going to write it up in an email.